File: 03. Stack and Queue/01. Valid parentheses/main.java
import java.util.Stack;

public class main {
    public static boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();

        for (char c : s.toCharArray()) {
            if (c == '(' || c == '{' || c == '[') {
                stack.push(c); // Push opening brackets
            } else {
                if (stack.isEmpty()) return false; // No matching opening bracket

                char top = stack.peek();
                if ((c == ')' && top == '(') || 
                    (c == '}' && top == '{') || 
                    (c == ']' && top == '[')) {
                    stack.pop(); // Valid pair, remove from stack
                } else {
                    return false; // Mismatch
                }
            }
        }
        return stack.isEmpty(); // Stack should be empty if all brackets match
    }

    public static void main(String[] args) {
        String[] testCases = {"()", "()[]{}", "(]", "([)]", "{[]}"};

        for (String test : testCases) {
            System.out.println("String: " + test + " -> " + (isValid(test) ? "Valid" : "Invalid"));
        }
    }
}

File: 03. Stack and Queue/01. Valid parentheses/main.py
def isValid(s):
    stack = []

    for c in s:
        if c in "({[":
            stack.append(c)  # Push opening brackets
        else:
            if not stack:  # No matching opening bracket
                return False
            
            top = stack[-1]
            if (c == ')' and top == '(') or (c == '}' and top == '{') or (c == ']' and top == '['):
                stack.pop()  # Valid pair, remove from stack
            else:
                return False  # Mismatch

    return not stack  # Stack should be empty if all brackets match

def main():
    testCases = ["()", "()[]{}", "(]", "([)]", "{[]}"]
    for test in testCases:
        print(f"String: {test} -> {'Valid' if isValid(test) else 'Invalid'}")

main()